cmd/clusterctl/client/alpha/rollout_rollbacker.go

```diff
@@ -0,0 +1,171 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package alpha
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/client/cluster"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/internal/util"
+	logf "sigs.k8s.io/cluster-api/cmd/clusterctl/log"
+	"sigs.k8s.io/cluster-api/controllers/mdutil"
+	"sigs.k8s.io/cluster-api/util/patch"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ObjectRollbacker will issue a rollback on the specified cluster-api resource.
+func (r *rollout) ObjectRollbacker(proxy cluster.Proxy, tuple util.ResourceTuple, namespace string, toRevision int64) error {
+	switch tuple.Resource {
+	case machineDeployment:
+		deployment, err := getMachineDeployment(proxy, tuple.Name, namespace)
+		if err != nil || deployment == nil {
+			return errors.Wrapf(err, "failed to fetch %v/%v", tuple.Resource, tuple.Name)
+		}
+		if deployment.Spec.Paused {
+			return errors.Errorf("can't rollback a paused MachineDeployment (run rollout resume first): %v/%v\n", tuple.Resource, tuple.Name)
+		}
+		if err := rollbackMachineDeployment(proxy, deployment, tuple.Name, namespace, toRevision); err != nil {
+			return err
+		}
+	default:
+		return errors.Errorf("invalid resource type %q, valid values are %v", tuple.Resource, validResourceTypes)
+	}
+	return nil
+}
+
+// rollbackMachineDeployment will rollback to a previous MachineSet revision used by this MachineDeployment.
+func rollbackMachineDeployment(proxy cluster.Proxy, d *clusterv1.MachineDeployment, name, namespace string, toRevision int64) error {
+	log := logf.Log
+	c, err := proxy.NewClient()
+	if err != nil {
+		return err
+	}
+
+	if toRevision < 0 {
+		return errors.Errorf("revision number cannot be negative: %v", toRevision)
+	}
+	msList, err := getMachineSetsForDeployment(proxy, d)
+	if err != nil {
+		return err
+	}
+	log.V(7).Info("Found MachineSets", "count", len(msList))
+	msForRevision, err := findMachineDeploymentRevision(toRevision, msList)
+	if err != nil {
+		return err
+	}
+	log.V(7).Info("Found revision", "revision", msForRevision)
+	patchHelper, err := patch.NewHelper(d, c)
+	if err != nil {
+		return err
+	}
+	// Copy template into the machinedeployment (excluding the hash)
+	revMSTemplate := *msForRevision.Spec.Template.DeepCopy()
+	delete(revMSTemplate.Labels, mdutil.DefaultMachineDeploymentUniqueLabelKey)
+
+	d.Spec.Template = revMSTemplate
+	return patchHelper.Patch(context.TODO(), d)
```

Comment:
I think we should avoid to use context.TODO but instead pass a context down the call chain or use a global ctx var, so, if required, if will be easier to set timeouts or cancelation.

Reply:
I will make a note of this and put it in a cleanup PR. It impacts several files outside of just `undo`. 
I will add this cleanup PR as a task to the tracking issue as well.

Reply:
Please open an issue to track this

Reply:
See: https://github.com/kubernetes-sigs/cluster-api/issues/4266